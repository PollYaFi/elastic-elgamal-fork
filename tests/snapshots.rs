//! Snapshot testing to check compatibility of `serde` serialization of types.

use insta::assert_yaml_snapshot;
use rand_chacha::ChaChaRng;
use rand_core::SeedableRng;

use elastic_elgamal::{
    app::{ChoiceParams, EncryptedChoice},
    group::{Generic, Group, Ristretto},
    Keypair, RangeDecomposition,
};

trait Named {
    const NAME: &'static str;
}

impl Named for Ristretto {
    const NAME: &'static str = "ristretto";
}

impl Named for Generic<::k256::Secp256k1> {
    const NAME: &'static str = "k256";
}

fn test_ciphertext_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();
    let ciphertext = public_key.encrypt(42, &mut rng);

    let full_name = format!("ciphertext-{}", G::NAME);
    assert_yaml_snapshot!(full_name, ciphertext);
}

fn test_zero_encryption_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();

    let (ciphertext, proof) = public_key.encrypt_zero(&mut rng);
    let ciphertext_with_proof = serde_json::json!({
        "ciphertext": ciphertext,
        "proof": proof,
    });
    let full_name = format!("zero-encryption-{}", G::NAME);
    assert_yaml_snapshot!(full_name, ciphertext_with_proof);
}

fn test_bool_encryption_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();

    let (ciphertext, proof) = public_key.encrypt_bool(true, &mut rng);
    let ciphertext_with_proof = serde_json::json!({
        "ciphertext": ciphertext,
        "proof": proof,
    });
    let full_name = format!("bool-encryption-{}", G::NAME);
    assert_yaml_snapshot!(full_name, ciphertext_with_proof);
}

fn test_range_encryption_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();
    let range = RangeDecomposition::optimal(100).into();

    let (ciphertext, proof) = public_key.encrypt_range(&range, 42, &mut rng);
    let ciphertext_with_proof = serde_json::json!({
        "ciphertext": ciphertext,
        "proof": proof,
    });
    let full_name = format!("range-encryption-{}", G::NAME);
    assert_yaml_snapshot!(full_name, ciphertext_with_proof);
}

fn test_encrypted_choice_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();
    let choice_params = ChoiceParams::single(public_key, 5);

    let choice = EncryptedChoice::single(3, &choice_params, &mut rng);
    let full_name = format!("encrypted-choice-{}", G::NAME);
    assert_yaml_snapshot!(full_name, choice);
}

fn test_encrypted_multi_choice_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();
    let choice_params = ChoiceParams::multi(public_key, 5);

    let choices = [false, true, true, false, true];
    let choices = EncryptedChoice::new(&choice_params, &choices, &mut rng);
    let full_name = format!("encrypted-multi-choice-{}", G::NAME);
    assert_yaml_snapshot!(full_name, choices);
}

mod ristretto {
    use super::*;

    #[test]
    fn ciphertext_snapshot() {
        test_ciphertext_snapshot::<Ristretto>();
    }

    #[test]
    fn zero_encryption_snapshot() {
        test_zero_encryption_snapshot::<Ristretto>();
    }

    #[test]
    fn bool_encryption_snapshot() {
        test_bool_encryption_snapshot::<Ristretto>();
    }

    #[test]
    fn range_encryption_snapshot() {
        test_range_encryption_snapshot::<Ristretto>();
    }

    #[test]
    fn encrypted_choice_snapshot() {
        test_encrypted_choice_snapshot::<Ristretto>();
    }

    #[test]
    fn encrypted_multi_choice_snapshot() {
        test_encrypted_multi_choice_snapshot::<Ristretto>();
    }
}

mod k256 {
    use super::*;

    type K256 = Generic<::k256::Secp256k1>;

    #[test]
    fn ciphertext_snapshot() {
        test_ciphertext_snapshot::<K256>();
    }

    #[test]
    fn zero_encryption_snapshot() {
        test_zero_encryption_snapshot::<K256>();
    }

    #[test]
    fn bool_encryption_snapshot() {
        test_bool_encryption_snapshot::<K256>();
    }

    #[test]
    fn range_encryption_snapshot() {
        test_range_encryption_snapshot::<K256>();
    }

    #[test]
    fn encrypted_choice_snapshot() {
        test_encrypted_choice_snapshot::<K256>();
    }

    #[test]
    fn encrypted_multi_choice_snapshot() {
        test_encrypted_multi_choice_snapshot::<K256>();
    }
}
