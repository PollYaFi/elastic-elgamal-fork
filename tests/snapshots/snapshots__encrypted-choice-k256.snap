---
source: tests/snapshots.rs
expression: choice

---
variants:
  - random_element: AgQbzI7_NulX5AZav0Ty90yRPsu9APHrKvYGo03s3Dn_
    blinded_element: Ap9uGRrxTZpokI5bSjq33H5WMsM6EwjF1o--egduVmgM
  - random_element: ArTsnXyCGVQRsyWMha-strO2-d3lZjFcnhysFN6qdz5F
    blinded_element: As6fY_BbuFwa1s8U4gQnrewXVrq0_Peuso4p2FrC3FQr
  - random_element: AvyDBYSVQ3AZO2Z3t9yOdQ_LEqhvwGwPCv6bL6OYuz87
    blinded_element: Aj1nlVV5A3sQzS-Z1pCk2LMkBBDjX_FJvxDiswYqfhSx
  - random_element: A1ccbaZpQHth3doN8JtL7UUKoSE024Bu637aUjdQcZKX
    blinded_element: ArByKTDJ1_ZoTHy6jyDCg-YWuOuvXN7cO5mRblNQbPFn
  - random_element: Az7PQJdBQlY3kkrJarOBrCSvmHmnjjJmL4NULjhnkydZ
    blinded_element: ApHfsPbl-nRo1vmXsFweeNo01ihSBNwm3Y59Nv_OTmmI
range_proof:
  common_challenge: YbuE2s-7bIM4B-j4kMzq1WNjri7dOr2YbuwM48IcQsA
  ring_responses:
    - cCaCg76iaRAVWthy6G3JqG4EOw1Dw7p_HgBfZ8kGWMo
    - A5GmkLDdIK_ElgFwnTy7KW9tnQxDslStKdohL-Lw-jo
    - APPJtwj4HeqzFoyIUSDxnEpUBtnKF8b_1-ETwEZON6Q
    - JOSPSDFwLEmTWdaVRhRgS_NBxHSMQyiPZC7DdBf2qz8
    - K2FeXUbwKbSVLPYj3zk2WOriHPk0wUeaQxKDikM6Hak
    - abjbejMVQqsxThw2mDuheunKgQFXOyTZPE8xXnj3J4Q
    - Rc6YO1oSKrOVLCUJsDvVJlA-isUazfUlNwwnvRPEFAI
    - wKI1juuh2eJ89djhTcbppYZsQuR3gdROj75GdLlR5do
    - LDK2bisgsAUSMDREARdQ8hFaxfw5CMDZOy4wsGSsziI
    - AtxlAGMPfUxBlUGZd1QEkqsdVx9foXiJEJ2RkN0uY3U
sum_proof:
  challenge: tG-4j34f5AR8xt7tWx5nmTYk4oLQgqExleWeyoO7Ze8
  response: zIg415cW6Emy9dD174-3DMslPiHNPkXqfdmHYbRpB5M

